Spearhead & Seedlet: Re-sort stylesheets for proper variable cascading.

allancole opened the pull request:
<!-- Thanks for contributing to our free themes! Please provide as much information as possible with your Pull Request by filling out the following - this helps make reviewing much quicker! -->

#### Changes proposed in this Pull Request:

- Seedlet: Enqueue an empty stylesheet to decouple parent styles and the CSS variables.
- Spearhead: Re-sort the empty stylesheet to allows parent styles/variables and child CSS-variables to cascade properly.
- Makes the custom color options work properly on the frontend and customizer.

Before|After
------------|------------
![2020-11-05 12 28 47](https://user-images.githubusercontent.com/709581/98275305-809d9880-1f62-11eb-8769-be446e3ecca5.gif)|![2020-11-05 12 26 23](https://user-images.githubusercontent.com/709581/98275048-30bed180-1f62-11eb-96b2-c18b9295f81a.gif)

#### To Do: 

- [ ] Replace the Seedlet default colors with the Spearhead defaults. 

See here: 
https://github.com/Automattic/themes/blob/fb71c3713afe7cd962cb0bdb62e82c672a82d5bb/seedlet/classes/class-seedlet-custom-colors.php#L23-L30

We may need to make this array in the Seedlet custom-colors class filterable by child themes. I have no idea if that’s actually possible though. 

#### Related issue(s):

#2701

allancole committed:
Spearhead: Resort the parent stylesheet
- Allows parent and child CSS-variables to cascade properly.

allancole committed:
Seedlet: Enqueue an empty stylesheet to decouple parent styles and variables.
- Adds a new empty stylesheet for the variable overrides.
- Then, resort variable overrides in the child theme.

allancole committed:
Spearhead: code tidying.

allancole commented:
I found a tricky issue with this solution thats related to the _To Do_ note above. On the frontend, any color that I don’t change in the customizer, resolve back to the defaults set in Spearhead instead of the colors actually set in the customizer in some cases. 

Customizer|Frontend
------------|------------
![image](https://user-images.githubusercontent.com/709581/98278683-f4da3b00-1f66-11eb-99d4-297b2a5de255.png)|![image](https://user-images.githubusercontent.com/709581/98278637-e2600180-1f66-11eb-8ff1-3172b7ace693.png)

This is happening because on the frontend, we’re doing a check to see if the custom color setting matches the default color settings before outputting the overrides. It only adds the CSS-variables override when it _doesn’t_ match the defaults set by Seedlet (not Spearhead). 

However, since Spearhead is a child theme and the color defaults are still hinged to the parent theme [here](https://github.com/Automattic/themes/blob/fb71c3713afe7cd962cb0bdb62e82c672a82d5bb/seedlet/classes/class-seedlet-custom-colors.php#L23-L30), the conditional check is checking against Seedlet’s defaults instead of Spearhead’s which means that _some_ Spearhead defaults won’t get overridden unless a user changes all the custom color options to not match the defaults in Seedlet. Kind of confusing to grasp, I know. 

The solution is to figure out a way for child themes to redefine the color option defaults that are set in the parent color options class here: 
https://github.com/Automattic/themes/blob/fb71c3713afe7cd962cb0bdb62e82c672a82d5bb/seedlet/classes/class-seedlet-custom-colors.php#L23-L30

MaggieCabrera commented:
I also see that when we change colors on the customizer after having used them in the editor, the editor is not showing the correct colors:

Customizer | Editor | Frontend
--- | --- | ---
<img width="1450" alt="Screenshot 2020-11-06 at 09 51 46" src="https://user-images.githubusercontent.com/3593343/98346053-c0718800-2015-11eb-8b62-fdd2d40f729a.png"> | <img width="935" alt="Screenshot 2020-11-06 at 09 51 52" src="https://user-images.githubusercontent.com/3593343/98346051-bfd8f180-2015-11eb-872f-8dd09d8a3d56.png"> | <img width="1023" alt="Screenshot 2020-11-06 at 09 51 56" src="https://user-images.githubusercontent.com/3593343/98346041-bd769780-2015-11eb-953c-77124383b2a5.png">

allancole committed:
Seedlet: Add custom variables for custom colors.
- Allows child themes to replace the parent themes defaults with their own.
- Only start custom colors class if the parent theme is active.
- Code tidying.

allancole committed:
Spearhead: Replace parent theme custom color defaults with spearheads defaults.

allancole committed:
Spearhead: Make sure custom color settings appear in the editor and frontend.

allancole committed:
Spearhead: Code tidying.

allancole commented:
So I think this PR is good to go but there are a couple of things that need a deeper look before merging.

### Seedlet doesn’t support darkmode but Spearhead does.

When setting custom colors, the darkmode rules get overwritten by the custom colors which effectively disable darkmode without a warning. We’ll probably want to add a warning about this to prevent confusion (maybe in a separate PR?). 

This probably doesn’t have to happen now or anytime soon, but the darkmode solution in Spearhead should ideally live in Seedlet. This comes with two benefits:

1. Gives us an easy way to support darkmode in potentially all Seedlet child themes.
2. With a more robust solution that lives in Seedlet, we could have options for default, dark and/or light mode custom colors in the customizer.

### Some Spearhead snippets that would need to be added to all Seedlet child-themes

**Connect editor palette colors to child theme’s CSS-variables / theme options:**

https://github.com/Automattic/themes/blob/ba85f09d19dbfa0226407ea1db17e8b9d86833b8/spearhead/functions.php#L78-L116

**Pass the child-theme’s color variables to the Seedlet custom colors class:**

https://github.com/Automattic/themes/blob/ba85f09d19dbfa0226407ea1db17e8b9d86833b8/spearhead/functions.php#L121-L131

### Seedlet’s inline style get reorganized.

**Re-organize seedlet CSS-variable files to respect cascading properly**

https://github.com/Automattic/themes/blob/ba85f09d19dbfa0226407ea1db17e8b9d86833b8/spearhead/functions.php#L150

https://github.com/Automattic/themes/blob/ba85f09d19dbfa0226407ea1db17e8b9d86833b8/spearhead/functions.php#L159-L160

Maybe we can do this in Seedlet itself? 

Also, the inline color variable overrides are no longer attached to the main Seedlet style sheet. Instead an empty stylesheet is registered and the inline color variable overrides get attached to that, which make it easier to resort without causing regressions in the variable cascading.

https://github.com/Automattic/themes/blob/ba85f09d19dbfa0226407ea1db17e8b9d86833b8/seedlet/classes/class-seedlet-custom-colors.php#L277-L286

jffng commented on seedlet/classes/class-seedlet-custom-colors.php:
```diff
@@ -263,7 +279,9 @@ function seedlet_generate_custom_color_variables( $context = null ) {
 	 */
 	function seedlet_custom_color_variables() {
```
Unrelated to this PR, but I think this function should be renamed to be more specific to what it does, since there is already a private variable of this class called `seedlet_custom_color_variables`. 

```suggestion
	function seedlet_enqueue_custom_color_variables() {
```

allancole replied:
Agreed. 

I kind of lean towards just doing it in this PR unless there’s a good reason not to.

jffng commented on seedlet/classes/class-seedlet-custom-colors.php:
```diff
@@ -263,7 +279,9 @@ function seedlet_generate_custom_color_variables( $context = null ) {
 	function seedlet_custom_color_variables() {
 		if ( 'default' !== get_theme_mod( 'custom_colors_active' ) ) {
-			wp_add_inline_style( 'seedlet-style', $this->seedlet_generate_custom_color_variables() );
+			wp_register_style( 'seedlet-variable-overrides', false );
```
Do we need to do the same re-ordering for enqueuing the editor custom colors?

allancole replied:
That’s a good question! It’s seems like we _don’t_ because editor colors are already working as expected on this PR, but still worth a look anyway.

allancole replied:
Yeah, so it looks like [this file](https://github.com/Automattic/themes/blob/fix/spearhead-custom-colors/seedlet/assets/css/custom-color-overrides.css) is completely useless, lol. We can go ahead and borrow the same enqueuing technique for the editor. Good catch!

jffng commented on seedlet/classes/class-seedlet-custom-colors.php:
```diff
@@ -14,19 +14,35 @@
+	private $default_border                 = '#EFEFEF';
 
-	function __construct() {
+	function __construct( $background, $foreground, $primary, $secondary, $tertiary, $border ) {
```
I think this needs to be:

```suggestion
	function __construct( $background = $default_background, $foreground = $default_foreground, $primary = $default_primary, $secondary = $default_secondary, $tertiary = $default_tertiary, $border = $default_border ) {
```

jffng reviewed (commented):
Thanks @allancole for a thorough explanation!

Re: moving dark mode to Seedlet, agree that it should be a different exploration.

> Some Spearhead snippets that would need to be added to all Seedlet child-themes

I'm comfortable with the snippets that need to be added to each child theme; they seem necessary at least until global styles replaces this mechanism. 
 
> Seedlet’s inline style get reorganized.

I also can't find a quick way around this, but I still am a bit confused as to why the stylesheets aren't already appended in the correct order.

allancole commented:
> I'm comfortable with the snippets that need to be added to each child theme; they seem necessary at least until global styles replaces this mechanism.

Yeah, thats probably the better way forward, all things considered.

allancole committed:
Update seedlet/classes/class-seedlet-custom-colors.php

Co-authored-by: Jeff Ong <jeff.ong@automattic.com>

allancole committed:
Spearhead: Rename stylesheet override enqueuing function.
- Function names should not match variable names.

allancole committed:
Merge branch 'fix/spearhead-custom-colors' of https://github.com/Automattic/themes into fix/spearhead-custom-colors

allancole committed:
Seedlet: Fix PHP in construct variables and reset default colors using the new method.

allancole committed:
Seedlet: Use an empty stylesheet to enqueue inline styles.
- Remove obsolete custom-color-overrides.css file.

scruffian commented:
The issue seems to be fixed by https://github.com/Automattic/themes/pull/3026. Please reopen if you think there's more we need to do.
